File: Leetcode 1301-1400/1390/1390.py
# Leetcode 1390: Four Divisors
# https://leetcode.com/problems/four-divisors/
# Solved on 4th of January, 2026
import math


class Solution:
    def sumFourDivisors(self, nums: list[int]) -> int:
        """
        Calculates the sum of divisors for numbers in the input list that have exactly four divisors.

        Args:
            nums (list[int]): A list of integers.
        Returns:
            int: The total sum of divisors for numbers with exactly four divisors.
        """
        totalSum = 0
        for num in nums:
            divisors = set()
            for i in range(1, int(math.sqrt(num)) + 1):
                if num % i == 0:
                    divisors.add(i)
                    divisors.add(num // i)

                    if len(divisors) > 4:
                        break

            if len(divisors) == 4:
                totalSum += sum(divisors)

        return totalSum

File: Leetcode 1301-1400/1390/1390.java
// Leetcode 1390: Four Divisors
// https://leetcode.com/problems/four-divisors/
// Solved on 4th of January, 2026
class Solution {
    /**
     * Calculates the sum of divisors for numbers that have exactly four divisors.
     * @param nums An array of integers.
     * @return The total sum of divisors for all numbers in `nums` that have exactly four divisors.
     */
    public int sumForDivisors(int[] nums) {
        int totalSum = 0;
        for (int num : nums) {
            int sum = 0;
            int count = 0;
            for (int i = 1; i * i <= num; i++) {
                if (num % i == 0) {
                    if (i * i == num) {
                        count++;
                        sum += i;
                    } else {
                        count += 2;
                        sum += i + num / i;
                    }
                    if (count > 4) {
                        break;
                    }
                }
            }
            if (count == 4) {
                totalSum += sum;
            }
        }
        return totalSum;
    }
}